Add support for kernel metadata with operates_on=domain. (Towards #925)

arporter opened the pull request:
This PR is purely about adding parsing and validation support for the new "operates_on=domain" kernel metadata. It does not address the changes needed for PSy-layer generation.

arporter commented:
I've noticed that we had effectively duplicated the check on the kernel 'operates-on' for the stub-generator for the LFRic API so I've removed one of them (the one in gen_kernel_stub.py).

arporter commented:
Ready for first review from either Iva or Rupert.

rupertford reviewed (changes requested):
Look good. The main thing is that I think the user-guide needs updating. Comments inline. I also have a potentially silly question. Are we limiting the use of `domain` to i-first fields and if so, don't we need metadata to specify this?

arporter commented on doc/user_guide/dynamo0p3.rst:
```diff
@@ -256,8 +256,9 @@ basis/differential-basis functions required by the kernel are to be evaluated.
 
-Kernel metadata may specify that a Kernel performs a stencil operation
-on a field. Any such metadata must provide a stencil type. See the
+The metadata for a Kernel which operates on a cell-column may specify
```
Yes, I was torn between the two options. I've now added separate sections (for the metadata rules and the argument-list rules).

rupertford replied:
I still think there is too little explanation at the moment. The user will want to know when to use domain and cell_column and why they are used and why there are restrictions - if any. I also think it might be better to restructure at some point. It does not seem right to introduce all the metadata options and then say domain supports a subset. However, seeing as we're just adding metadata support, this can be left until the actual implementation.

arporter commented on src/psyclone/domain/lfric/lfric_arg_descriptor.py:
```diff
@@ -350,9 +350,13 @@ def _init_field(self, arg_type, operates_on):
                             passed to a kernel that operates on cell-columns \
                             does not have a valid access (one of [READ, INC]).
+        :raises ParseError: if the kernel operates on the domain and is \
+                            passed a field on a discontinuous space.
```
Oops/

arporter commented:
I've addressed your comments. At the moment we have no concept of an 'i-first' field so no, we're not restricting which fields can be passed to this new kernel type (other than the fact that they have to be on discontinuous spaces). When we come to trying to minimise the number of times an LFRic field is mapped to an i-first field then we will need to keep track. Currently, the mapping is not just data layout though - it involves some science intelligence too. I can't remember what we decided to do about that.
Ready for another look.

rupertford reviewed (approved):
All issues have been addressed. Approving.
